### ISSUE TITLE
Narrowing of generic types from a method does not work

### ISSUE THREAD

**Issue by pelme** [python/mypy#20898]
**To Reproduce**
Using a class with a generic type that is returned in a method as a `type[T]` does not work properly in isinstance narrowing checks:

https://mypy-play.net/?mypy=latest&python=3.12&gist=dd509c91ecf07303e62a00d2d6646345

**Expected Behavior**
mypy should narrow the union and no error should be emitted.

- ty does the right thing: https://play.ty.dev/2badea13-87aa-46d0-b415-28c98c54b6ac
- pyright has a similar issue to mypy

**Actual Behavior**
mypy does not narrow the type correctly:

```
main.py:27: error: Item "A" of "A | B" has no attribute "b_only"  [union-attr]
```

**Comment by A5rocks** [comment]
It would be valid for `get_type` to return a type that is a subtype of `A`, which means the type should still be `A | B`.

**Comment by pelme** [comment]
B is not a subtype of A? 

added reveal_type that shows that mypy does narrow it to A in the if clause but then "forgets" about it after the return:

https://mypy-play.net/?mypy=latest&python=3.12&gist=8f49bd1a57b7b28f72181038124ba563

**Comment by A5rocks** [comment]
This is about narrowing in the else branch, not the if. In this case, if get_type returned a subtype something that is an instance of A would go to the else branch and not return.

**Comment by pelme** [comment]
how is that different from this example where A is directly passed to the isinstance check(not via the generic type)? https://mypy-play.net/?mypy=latest&python=3.12&gist=d3b26e9007b44af8753c85d6eed2ea42

**Comment by A5rocks** [comment]
Because type[A] is covariant, so could be any subtype of A.

**Comment by A5rocks** [comment]
Maybe to make things more concrete:
```py
class A:
    ...
class B:
    ...

def f(x: A | B, y: type[A]):
    if isinstance(x, y):
        return
    
    checked: B = x  # obviously mypy shouldn't allow this
```

**Comment by carljm** [comment]
This was a bug in ty, it will be fixed by https://github.com/astral-sh/ruff/pull/23598 - thanks @pelme and @A5rocks for flagging it!

**Comment by pelme** [comment]
Thanks for looking into this and explaining my misunderstanding.

I have refactored my own code which triggered this to use another approach instead.

I am glad my confusion of this could be helpful and in the end improve ty. 😆
